url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XScriptEngine.java

```diff
@@ -0,0 +1,909 @@
+		return new XValueObj(pointer);
+	}
+
+	@Override
+	public XValue alloc(String string) {
+		return alloc(getBaseType(XUtils.STRING), string);
+	}
+
+	@Override
+	public XValue createTuple(List<XValue> list) {
+		return alloc(getBaseType(XUtils.TUPLE), list);
+	}
+
+	@Override
+	public XValue createTuple(XValue... args) {
+		List<XValue> list = new ArrayList<XValue>(args.length);
+		for (XValue value : args) {
+			list.add(value);
+		}
+		return createTuple(list);
+	}
+
+	@Override
+	public XValue createList(List<XValue> list) {
+		return alloc(getBaseType(XUtils.LIST), list);
+	}
+
+	@Override
+	public XValue createMap(Map<String, XValue> map) {
+		return alloc(getBaseType(XUtils.MAP), map);
+	}
+
+	@Override
+	public XFunctionData getFunction(String name) {
+		Bindings functions = (Bindings) get(XScriptLang.ENGINE_ATTR_FUNCTIONS_BINDING);
+		return (XFunctionData) functions.get(name);
+	}
+
+	@Override
+	public void addNativeMethod(String name, XFunctionData function) {
+		Bindings functions = (Bindings) get(XScriptLang.ENGINE_ATTR_FUNCTIONS_BINDING);
+		if (functions == null) {
+			put(XScriptLang.ENGINE_ATTR_FUNCTIONS_BINDING,
+					functions = createBindings());
+		}
+		functions.put(name, function);
+	}
+
+	@Override
+	public XValue createFunction(String name) {
+		return alloc(getBaseType(XUtils.NATIVE_FUNC), name);
+	}
+
+	@Override
+	public void gc() {
+		for (XObject[] m : memory) {
+			for (XObject o : m) {
+				o.resetVisible();
+			}
+		}
+		for (XValue baseType : baseTypes) {
+			baseType.setVisible(this);
+		}
+		for (XValue module : modules.values()) {
+			module.setVisible(this);
+		}
+		for (XExec exec : threads) {
+			exec.setVisible();
+		}
+		for (XObject[] m : memory) {
+			for (int i = 0; i < m.length; i++) {
+				if (!m[i].isVisible()) {
+					m[i].delete(this, false);
+				}
+			}
+		}
+		// TODO visited all the objects?
+	}
+
+	private Object writeReplace() throws ObjectStreamException {
+		return new Save(this);
+	}
+
+	private static class Save implements Serializable, Externalizable {
+
+		private static final long serialVersionUID = -7988390441980350547L;
+
+		private XScriptEngine scriptEngine;
+
+		@SuppressWarnings("unused")
+		Save() {
+			scriptEngine = new XScriptEngine();
+		}
+
+		Save(XScriptEngine scriptEngine) {
+			this.scriptEngine = scriptEngine;
+		}
+
+		private Object readResolve() throws ObjectStreamException {
+			return scriptEngine;
+		}
+
+		@Override
+		public void writeExternal(ObjectOutput out) throws IOException {
+			scriptEngine.writeExternal(out);
+		}
+
+		@Override
+		public void readExternal(ObjectInput in) throws IOException,
+				ClassNotFoundException {
+			scriptEngine.readExternal(in);
+		}
+
+	}
+
+	@Override
+	public long getTime() {
+		return System.currentTimeMillis();
+	}
+
+	@Override
+	public XValue getBuiltinModule() {
+		return getModule("__builtin__");
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public XConstPoolImpl loadModule(String name) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 29. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 813-825 into a separate method.